### ISSUE TITLE
Investigate and fix MySQL issues in -stage/-prod

### ISSUE THREAD

**Issue by willdurand** [mozilla/addons-server#15980]
We had a MySQL issue on -stage after having deployed this week's tag, see: https://github.com/mozilla/addons-server/issues/15975

There is an incompatibility between the current MySQL version running in -stage/-prod (5.7.26 at the time of writing) and the changes introduced in https://github.com/mozilla/addons-server/pull/15932 (including this commit: https://github.com/mozilla/addons-server/commit/374b3ec84e9a5ee0957d76b59680d6c7e2415138), which were needed for MySQL 8 (local dev/-dev).

We fixed the problem for this week's tag by reverting two commits on top of the initial tag. That means the `master` branch couldn't be deployed to -stage/-prod for the same reasons. We did not revert the two commits on `master` because -dev and local dev use MySQL 8 and it seems to work (+ rollback seems tricky for local dev).

**Comment by willdurand** [comment]
Note: I don't think we can deploy MySQL 8 on -stage/-prod to get around this problem because of https://github.com/mozilla/addons-server/issues/15970

**Comment by willdurand** [comment]
We found different charset values between our different envs and this is suspicious:

Local dev:

```
mysql> SHOW VARIABLES LIKE 'char%';
+--------------------------+--------------------------------+
| Variable_name            | Value                          |
+--------------------------+--------------------------------+
| character_set_client     | latin1                         |
| character_set_connection | latin1                         |
| character_set_database   | utf8mb4                        |
| character_set_filesystem | binary                         |
| character_set_results    | latin1                         |
| character_set_server     | utf8mb4                        |
| character_set_system     | utf8                           |
| character_sets_dir       | /usr/share/mysql-8.0/charsets/ |
+--------------------------+--------------------------------+
```

-dev:

```
mysql> SHOW VARIABLES LIKE 'char%';
+--------------------------+-------------------------------------------+
| Variable_name            | Value                                     |
+--------------------------+-------------------------------------------+
| character_set_client     | utf8mb4                                   |
| character_set_connection | utf8mb4                                   |
| character_set_database   | utf8mb4                                   |
| character_set_filesystem | binary                                    |
| character_set_results    | utf8mb4                                   |
| character_set_server     | utf8mb4                                   |
| character_set_system     | utf8                                      |
| character_sets_dir       | /rdsdbbin/mysql-8.0.21.R8/share/charsets/ |
+--------------------------+-------------------------------------------+
```

-stage:

```
MySQL [addons_allizom_org]> SHOW VARIABLES LIKE 'char%';
+--------------------------+-------------------------------------------+
| Variable_name            | Value                                     |
+--------------------------+-------------------------------------------+
| character_set_client     | utf8mb4                                   |
| character_set_connection | utf8mb4                                   |
| character_set_database   | utf8mb4                                   |
| character_set_filesystem | binary                                    |
| character_set_results    | utf8mb4                                   |
| character_set_server     | latin1                                    |
| character_set_system     | utf8                                      |
| character_sets_dir       | /rdsdbbin/mysql-5.7.26.R1/share/charsets/ |
+--------------------------+-------------------------------------------+
```

-prod:


```
MySQL [addons_mozilla_org]> SHOW VARIABLES LIKE 'char%';
+--------------------------+-------------------------------------------+
| Variable_name            | Value                                     |
+--------------------------+-------------------------------------------+
| character_set_client     | utf8mb4                                   |
| character_set_connection | utf8mb4                                   |
| character_set_database   | utf8mb4                                   |
| character_set_filesystem | binary                                    |
| character_set_results    | utf8mb4                                   |
| character_set_server     | utf8                                      |
| character_set_system     | utf8                                      |
| character_sets_dir       | /rdsdbbin/mysql-5.7.26.R1/share/charsets/ |
+--------------------------+-------------------------------------------+
```

The -dev config looks good, i.e. all charsets should be set to `utf8mb4` (that will evolve in the future but it is a separate issue).

**Comment by diox** [comment]
For the record, I agree: the dev config is what we want everywhere at the moment.

**Comment by willdurand** [comment]
@smarnach fixed the server charset on -stage and I just created this tag to see if it fixes this issue: https://github.com/mozilla/addons-server/releases/tag/2020.11.13

**Comment by willdurand** [comment]
mm -stage now shows:

```
show variables like 'char%';
+--------------------------+-------------------------------------------+
| Variable_name            | Value                                     |
+--------------------------+-------------------------------------------+
| character_set_client     | latin1                                    |
| character_set_connection | latin1                                    |
| character_set_database   | utf8mb4                                   |
| character_set_filesystem | binary                                    |
| character_set_results    | latin1                                    |
| character_set_server     | utf8mb4                                   |
| character_set_system     | utf8                                      |
| character_sets_dir       | /rdsdbbin/mysql-5.7.26.R1/share/charsets/ |
+--------------------------+-------------------------------------------+
```

This looks like local dev, not like -dev. Could it be that the new mysql client libs updated some of these values?

**Comment by willdurand** [comment]
Reconfiguring the charsets with `utf8mb4` fixes the problem on -stage. We still don't know why values have been changed but it works once the values are correct.

**Comment by willdurand** [comment]
I am going to create a new tag (2020.11.13-1) and deploy it to -stage to see if charset config remains unchanged.

**Comment by willdurand** [comment]
> I am going to create a new tag (2020.11.13-1) and deploy it to -stage to see if charset config remains unchanged.

After the deployment, the site remains functional and the mysql config hasn't been changed. This means we "simply" have to configure those charset variables before next week's push. cc @bqbn

**Comment by bqbn** [comment]
> > I am going to create a new tag (2020.11.13-1) and deploy it to -stage to see if charset config remains unchanged.
> 
> After the deployment, the site remains functional and the mysql config hasn't been changed. This means we "simply" have to configure those charset variables before next week's push. cc @bqbn

Could you help put what needs to be configured in next week's push doc please?

**Comment by willdurand** [comment]
> Could you help put what needs to be configured in next week's push doc please?

Yes, I will update the push doc

**Comment by bqbn** [comment]
@willdurand Thanks for the info. In the doc, it reads,

```
so it's important to re-run the SHOW VARIABLES LIKE 'char%';
```

What is the correct timing of re-running the SHOW command during push? For example, is it after the "set" command, before the db migration, after the db migration or some other times?

**Comment by willdurand** [comment]
> What is the correct timing of re-running the SHOW command during push? For example, is it after the "set" command, before the db migration, after the db migration or some other times?

I suppose once all mysql instances are updated, we should verify that the config is correct and then start the push. Given that most values are correct in -prod already, it shouldn't break anything because only `character_set_server` would change. @smarnach did it on -stage earlier today so let's see what he thinks

**Comment by bqbn** [comment]
After today's push, it was reported that smileys were not shown correctly. We had to change the default values for the following parameters in -prod (which runs v5.7) and restart the app instances to fix the issue.

* character_set_client
* character_set_connection
* character_set_results 

Before the push, we did verify that those parameters have the expected `utf8mb4` values inside an app container. However, after the push, inside an app container, we saw that the values of those parameters were `utf8`, but not `utf8mb4`. `utf8` is the v5.7 db engine's default value.

@willdurand mentioned that today's push contains the new mysql client libs. It may be that the new lib code doesn't pass those "character_set_*" parameters when it connects to the database, whereas the old client lib does. Thus, when the new client lib creates a session, the session uses the db engine's default values, which are `utf8`.

We have now set the db engines default values for those said parameters to be `utf8mb4` in both -stage and -prod environments, so the smiley-not-shown-correctly issue should not happen again.

Note that in MySQL v8.0, the db engine's default values for the mentioned "character_set_*" parameters are `utf8mb4` already. It is probably why we didn't experience the same issue in -dev environment (which runs v8.0). And when we upgrade -stage and -prod to v8.0, even if we don't explicitly set the default values for those parameters, the db engine's defaults are what we want.
